Components: Stop matching autocompleter upon mismatch

fullofcaffeine opened the pull request:
Attempts to fix https://github.com/WordPress/gutenberg/issues/30640.

## How to test

1. Spin up a `wp-env` instance in this branch, load the editor by creating a new post or page, open the JS console to check its output;
1. Write an user mention on the page (e.g "@ad... for "admin"), it should bring the autocompleter popup
1. Click any of the mention in the page, then start typing after it. Try typing a text that matches, it should work as expected and filter in the popup
1. Try typing a text that doesn't match anything, it should eventually run out of options
1. Keep typing, you shouldn't have any slowdown
1. Completion with spaces should work as well
1. If you backspace from unmatched text up to a portion that matches a name, it should bring the popup
1. Try the block completer (`/` in a new paragraph). It should work as expected and match blocks with spaces, too.

fullofcaffeine committed:
Stop matching upon mismatch, reset when trigger char or backspace is pressed

ockham commented:
Thanks for taking this on!

Unfortunately, it seems that with this PR, the suggestions box doesn't always appear for me when typing e.g. `@admin`. There might be a really long delay, but in effect, it seems like the box doesn't appear at all most times 😕 

Furthermore, one of the autocomplete related e2e tests is currently failing:

```
FAIL specs/editor/various/mentions.test.js (36.816 s)
  ● autocomplete mentions › should insert mention

    TimeoutError: waiting for selector `.components-autocomplete__result` failed: timeout 30000ms exceeded

      16 | 		await clickBlockAppender();
      17 | 		await page.keyboard.type( 'I am @a' );
    > 18 | 		await page.waitForSelector( '.components-autocomplete__result' );
         | 		           ^
      19 | 		await page.keyboard.press( 'Enter' );
      20 | 		await page.keyboard.type( '.' );
      21 | 		expect( await getEditedPostContent() ).toMatchSnapshot();
```

BTW maybe we can avoid introducing `mismatch` state, and instead rely on the existing `reset()` method (which is [effectively called](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L469-L472) if [no completer is found](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L438-L467) already, except for those [newly introduced whitespace exceptions](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L458-L461). I'm basically hoping that we can maybe tweak that logic a bit in order to stop looking for matches upon a mismatch 🙂

fullofcaffeine commented:
> Thanks for taking this on!
> 
> Unfortunately, it seems that with this PR, the suggestions box doesn't always appear for me when typing e.g. `@admin`. There might be a really long delay, but in effect, it seems like the box doesn't appear at all most times confused
> 
> Furthermore, one of the autocomplete related e2e tests is currently failing:
> 
> ```
> FAIL specs/editor/various/mentions.test.js (36.816 s)
>   ● autocomplete mentions › should insert mention
> 
>     TimeoutError: waiting for selector `.components-autocomplete__result` failed: timeout 30000ms exceeded
> 
>       16 | 		await clickBlockAppender();
>       17 | 		await page.keyboard.type( 'I am @a' );
>     > 18 | 		await page.waitForSelector( '.components-autocomplete__result' );
>          | 		           ^
>       19 | 		await page.keyboard.press( 'Enter' );
>       20 | 		await page.keyboard.type( '.' );
>       21 | 		expect( await getEditedPostContent() ).toMatchSnapshot();
> ```
> 
> BTW maybe we can avoid introducing `mismatch` state, and instead rely on the existing `reset()` method (which is [effectively called](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L469-L472) if [no completer is found](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L438-L467) already, except for those [newly introduced whitespace exceptions](https://github.com/WordPress/gutenberg/blob/33683710dd7694427b62a22f03f9134aaed50ae8/packages/components/src/autocomplete/index.js#L458-L461). I'm basically hoping that we can maybe tweak that logic a bit in order to stop looking for matches upon a mismatch slightly_smiling_face

Thanks for having a look! That's true, I managed to get into a state where the box isn't shown consistently :(. I'll spend a bit more time today seeing if I can fix this.

fullofcaffeine committed:
Simplify

ockham commented on packages/components/src/autocomplete/index.js:
```diff
@@ -471,12 +477,21 @@ function Autocomplete( {
+		// we didn't have a new trigger typed, then we should not continue with this effect.
+		const mismatch = filteredOptions.length === 0;
+		if (mismatch && text.slice(-1) !== completer.triggerPrefix) {
+			return;
```
Do we need to `reset()` prior to `return`ing? 🤔 

```suggestion
			reset();
			return;
```

fullofcaffeine replied:
This is not relevant anymore as I changed and simplified the implementation a bit.

fullofcaffeine committed:
Match and query only when relevant / needed

fullofcaffeine committed:
Remove backspacing logic as it is not needed

fullofcaffeine committed:
Rename triggerMatch to triggerMatchHint (since it might not contain the whole match, as it is split by space

fullofcaffeine commented on packages/components/src/autocomplete/index.js:
```diff
@@ -404,12 +405,10 @@ function Autocomplete( {
 			case ENTER:
 				select( filteredOptions[ selectedIndex ] );
 				break;
-
```
OOpsie

fullofcaffeine committed:
Tweak and additional debugging logs

fullofcaffeine committed:
Revert some unneeded line-break changes

fullofcaffeine committed:
Small cleanup

fullofcaffeine committed:
Simplify a bit more

fullofcaffeine committed:
Add the tooDistantFromTrigger guard again

fullofcaffeine committed:
Remove debug code and improve UX when backspacing

When backspacing, we were only considering a single word, which was less than ideal and would not trigger the effect if the user was in the middle of fixing a completion for a two-word name for example.

fullofcaffeine committed:
Fix typo in comment

fullofcaffeine committed:
Improve const name

gwwar commented on packages/components/src/autocomplete/index.js:
```diff
@@ -444,17 +455,54 @@ function Autocomplete( {
+				// significantly. This could happen, for example, if `matchingWhileBackspacing`
+				// is true and one of the "words" end up being too long. If that's the case,
+				// it will be caught by this guard.
+				if ( tooDistantFromTrigger ) return false;
```
From [WP guidelines](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#spacing)

> if/else/for/while/try blocks should always use braces

This is more of a guard against future changes. Basic idea on the why we prefer braces is that it's easy for folks to accidentally add a line and think it's part of the if block, when it's not.

I'm surprised the linter didn't autofix this.

fullofcaffeine replied:
Ok, thanks for the nudge!

gziolo replied:
It's controlled by Prettier, we can't do much about it. It has to stay as is.

gwwar replied:
Interesting, this option is a won't fix too: https://github.com/prettier/prettier/issues/4262

Do we think code style guideline docs could be simplified to lint checks pass (in pretter / phpcs)?

gziolo replied:
> I'm surprised the linter didn't autofix this.

All ESLint rules that conflict with Prettier formatting are disabled by the config they have for ESLint.

PHP and JS coding guidelines are different. I see that JS coding standards need to clarify that bit in https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#blocks-and-curly-braces

gziolo replied:
I added this topic to the Core JS weekly chat next week: https://docs.google.com/document/d/1Y2Q1iZ89M88NA3iq0TyLzWn9wy6Zxob-wX3tJ4ivHPc/edit#

gwwar commented on packages/components/src/autocomplete/index.js:
```diff
@@ -444,17 +455,54 @@ function Autocomplete( {
+					index + triggerPrefix.length
+				);
+
+				const tooDistantFromTrigger = textWithoutTrigger.length > 50; // 50 chars seems to be a good limit.
```
For folks reviewing the trigger is a key like `/` or `@` and `textWithoutTrigger` are all characters after that. So for something like `@someusernamewithfiftycharacters` it will still properly early return after 50 chars even if folks don't enter a space. 

If we're picking something arbitrary, max user_login length is [60 characters](https://codex.wordpress.org/Database_Description#Table:_wp_users). I don't feel strongly on this one.

fullofcaffeine replied:
Yeah, it was arbitrary. I can adjust it to 60 if we think it's better. I don't feel too happy about this guard, but it's needed for the case described as long as we use this early return approach.

gwwar replied:
I'm okay either way on this one 👍

gwwar reviewed (approved):
Thanks @fullofcaffeine, I'm giving tentative approval. We'll want a +1 for which path to take, as I haven't interacted much with this code area before. Either we take this patch and follow up on the approach, or perhaps consider reverting 29939 or making it exclusive to the block inserter auto-complete as @mtias suggested.

Overall, I think this works well as an interim patch to address the performance issue added in https://github.com/WordPress/gutenberg/pull/29939. The approach of exiting early when potential completion text is too long is easy to reason about.

If we want to test other approaches in the future, I'd highly recommend adding unit tests (but not component tests). It might take a bit of refactoring, but if we could isolate the completer function, being able to pass in some content, and have it return whether or not it's a match will help document all the cases we want to keep, and make it easier to try other approaches to speeding this up.

We already have some basic E2E coverage. I don't think we need more unless specific things keep breaking that unit tests cannot catch.

Some quick 🤔 ideas to experiment with in follow-ups:

- We're currently running this on every keypress. When a user is typing rapidly we probably want to get out of the way. I suspect folks will slow down on typing speed when trying to autocomplete. Tuning a debounce would likely make things feel more lively.
- In a similar vein should we store more state to avoid running this whenever content changes? (Memo, detect a trigger, stop when a substring has no matches etc). The tradeoff here is that the more complex our approach is, the more likely we'll be wrong. We'll also want to be careful about making something slower accidently so we'll need to perf profile whatever we experiment with.

youknowriad commented on packages/components/src/autocomplete/index.js:
```diff
@@ -444,17 +455,54 @@ function Autocomplete( {
+				// Ex: "Some text @marcelo sekkkk" <--- "kkkk" caused a mismatch, but
+				// if the user presses backspace here, it will show the completion popup again.
+				const matchingWhileBackspacing =
+					backspacing && textWithoutTrigger.split( /\s/ ).length <= 3;
```
As suggested by @gwwar it's not very clear what these checks do, so yeah extracting the function + unit tests sounds like a great approach.

The release need to happen soon, I might merge as is though.

gziolo commented:
I would personally go with the bug fix and observe how it works. We need to make this autocomplete feature useful with all the block patterns and variations that are hard to match with only one keyword.

The approach proposed is close to what was discussed in other issues/PRs on GitHub.

We definitely should have another iteration before WP 5.8 to improve it further.

gziolo commented:
@fullofcaffeine, thank you for all your work to improve the autocomplete feature. Are you going to work on a follow-up or should we file an issue with all details?

fullofcaffeine commented:
> @fullofcaffeine, thank you for all your work to improve the autocomplete feature. 

Thanks! ❤️  No problem! Glad I could help! :)

> Are you going to work on a follow-up or should we file an issue with all details?

I don't plan on working on a follow-up just now, I'll let it cool-down for a while, and might resume it in the future if someone else more experienced with the code-base doesn't beat me to it. If you could create a follow-up issue, it'd be great! cc @gwwar

gwwar commented:
Added a follow up issue in https://github.com/WordPress/gutenberg/issues/30969
